File: apps/mobile/client-apple/VettaKit/Sources/VettaKit/Protocol/RemoteConnection.swift
import Foundation

public enum RemoteConnectionEvent {
	case state(RemoteConnectionState)
	case remoteRequest(RemoteRequest)
	case remoteEvent(RemoteEvent)
	case peerStatus(online: Bool)
	case error(RemoteError)
}

public enum RemoteHelloDecision {
	case approve
	/// Ask the person at the acceptor to compare the verification code first.
	case pending(approval: () async -> Bool)
	case reject(reason: String)
}

public struct RemoteConnectionSnapshot: Equatable {
	public var state: RemoteConnectionState
	public var deviceId: String
	public var connectionId: String
	public var peerDeviceId: String?
	public var peerIdentityKey: String?
	/// Six-digit code both ends can display to confirm a manual pairing.
	public var verificationCode: String?
	public var lastEventSequence: Int
	public var lastAckSequence: Int
	public var pendingRequestCount: Int
	public var reconnectCount: Int
	public var lastRttMs: Double?
	public var lastErrorCode: RemoteErrorCode?
}

public struct RemoteConnectionOptions {
	public var role: RemoteRole
	public var deviceId: String
	public var deviceName: String
	public var capabilities: RemoteCapabilities
	public var identity: RemoteIdentityKeyPair
	/// `initiate` sends `hello`; `accept` waits for the peer's `hello` (the desktop LAN server).
	public var handshake: Handshake = .initiate
	/// Pinned peer identity; a peer presenting another key is rejected before any secret is derived.
	public var expectedPeerIdentityKey: Data?
	/// Acceptor-only hook that decides whether an unknown peer may pair.
	public var onHello: ((RemoteHello) async -> RemoteHelloDecision)?
	/// Shared outbound journal so events survive a channel switch.
	public var journal: RemoteEventJournal?
	/// Highest event sequence already received from this peer.
	public var resumeFrom: Int = 0
	public var connectionId: String?
	public var requestTimeoutMs: Double = 30_000
	public var now: () -> Double = WallClock.nowMs
	public var randomBytes: RandomBytes = RemoteCrypto.randomBytes

	public enum Handshake { case initiate, accept }

	public init(role: RemoteRole, deviceId: String, deviceName: String, capabilities: RemoteCapabilities, identity: RemoteIdentityKeyPair) {
		self.role = role
		self.deviceId = deviceId
		self.deviceName = deviceName
		self.capabilities = capabilities
		self.identity = identity
	}
}

public struct RemoteRequestError: Error, LocalizedError, Equatable {
	public let message: String
	public init(_ message: String) { self.message = message }
	public var errorDescription: String? { message }
}

/// One end of a remote link over one transport (port of `connection.ts`). Handles
/// the plaintext handshake, derives the session keys, seals everything after
/// that, correlates requests and keeps the inbound event sequence continuous.
public final class RemoteConnection {
	private struct PendingRequest {
		let continuation: CheckedContinuation<JSONValue?, Error>
		let startedAt: Double
		let timeout: Task<Void, Never>
	}

	private static let earlySealedLimit = 32

	public private(set) var state: RemoteConnectionState = .idle
	private let transport: RemoteTransport
	private let options: RemoteConnectionOptions
	private let connectionId: String
	private let journal: RemoteEventJournal
	private var pending: [String: PendingRequest] = [:]
	private let listeners = Listeners<RemoteConnectionEvent>()
	private var ephemeral: RemoteIdentityKeyPair?
	private var keys: RemoteSessionKeys?
	private var peerDeviceId: String?
	private var peerIdentityKey: Data?
	private var lastEventSequence: Int
	private var lastAckSequence = 0
	private var reconnectCount = 0
	private var lastRttMs: Double?
	private var lastErrorCode: RemoteErrorCode?
	private var requestCounter = 0
	/// Sealed frames that overtook our own `hello_ack` on a relay; held instead of failing.
	private var earlySealed: [RemoteSealed] = []

	public init(transport: RemoteTransport, options: RemoteConnectionOptions) {
		self.transport = transport
		self.options = options
		connectionId = options.connectionId ?? "conn-\(options.deviceId)-\(RemoteConnection.shortRandom())"
		journal = options.journal ?? RemoteEventJournal(now: options.now)
		lastEventSequence = options.resumeFrom
	}

	private static func shortRandom() -> String {
		let alphabet = Array("abcdefghijklmnopqrstuvwxyz0123456789")
		return String((0 ..< 8).map { _ in alphabet.randomElement()! })
	}

	@discardableResult
	public func onEvent(_ listener: @escaping (RemoteConnectionEvent) -> Void) -> () -> Void {
		listeners.add(listener)
	}

	public var snapshot: RemoteConnectionSnapshot {
		RemoteConnectionSnapshot(
			state: state,
			deviceId: options.deviceId,
			connectionId: connectionId,
			peerDeviceId: peerDeviceId,
			peerIdentityKey: peerIdentityKey.map(Base64URL.encode),
			verificationCode: peerIdentityKey.map { RemoteCrypto.verificationCode(options.identity.publicKey, $0) },
			lastEventSequence: lastEventSequence,
			lastAckSequence: lastAckSequence,
			pendingRequestCount: pending.count,
			reconnectCount: reconnectCount,
			lastRttMs: lastRttMs,
			lastErrorCode: lastErrorCode
		)
	}

	public func connect() async throws {
		if state == .online || state == .connecting || state == .pendingApproval { return }
		setState(state == .idle ? .connecting : .reconnecting)
		ephemeral = RemoteIdentityKeyPair.generate(randomBytes: options.randomBytes)
		keys = nil
		earlySealed.removeAll()
		do {
			try await transport.connect(RemoteTransportHandlers(
				onFrame: { [weak self] frame in self?.handleFrame(frame) },
				onClose: { [weak self] reason in self?.handleClose(reason) }
			))
			if options.handshake == .initiate { try transport.send(buildHello()) }
		} catch {
			setState(.failed)
			throw error
		}
	}

	public func close() {
		if state == .closed { return }
		rejectPending("remote connection closed")
		setState(.closed)
		keys = nil
		transport.close(reason: nil)
	}

	public func request(_ method: RemoteRequestMethod, payload: JSONValue? = nil, sessionId: String? = nil) async throws -> JSONValue? {
		guard state == .online else { throw RemoteRequestError("remote connection is \(state.rawValue)") }
		let requestId = "\(connectionId)-\(Int64(options.now()))-\(requestCounter)"
		requestCounter += 1
		let request = RemoteRequest(requestId: requestId, method: method, sessionId: sessionId, payload: payload)
		return try await withCheckedThrowingContinuation { continuation in
			let timeout = schedule(after: options.requestTimeoutMs) { [weak self] in
				guard let self, let entry = self.pending.removeValue(forKey: requestId) else { return }
				let message = "remote request timed out: \(method.rawValue)"
				self.lastErrorCode = .requestTimeout
				self.emit(.error(RemoteError(code: .requestTimeout, message: message, retryable: true)))
				entry.continuation.resume(throwing: RemoteRequestError(message))
			}
			pending[requestId] = PendingRequest(continuation: continuation, startedAt: options.now(), timeout: timeout)
			do {
				try sendSealed(.request(request))
			} catch {
				timeout.cancel()
				pending.removeValue(forKey: requestId)
				continuation.resume(throwing: error)
			}
		}
	}

	public func respond(requestId: String, success: Bool, payload: JSONValue? = nil, error: RemoteError? = nil) throws {
		guard state == .online else { throw RemoteRequestError("remote connection is \(state.rawValue)") }
		try sendSealed(.response(RemoteResponse(requestId: requestId, success: success, payload: payload, error: error)))
	}

	/// Records the event in the journal and delivers it when online.
	@discardableResult
	public func emitEvent(_ name: RemoteEventName, payload: JSONValue? = nil, sessionId: String? = nil) throws -> RemoteEvent {
		let sequence = journal.nextSequence()
		let event = RemoteEvent(eventId: "\(options.deviceId)-event-\(sequence)", sequence: sequence, name: name, sessionId: sessionId, payload: payload)
		journal.remember(event)
		if state == .online { try sendSealed(.event(event)) }
		return event
	}

	private func buildHello() -> RemoteFrame {
		.hello(RemoteHello(
			role: options.role,
			deviceId: options.deviceId,
			deviceName: options.deviceName,
			capabilities: options.capabilities,
			connectionId: connectionId,
			identityKey: Base64URL.encode(options.identity.publicKey),
			ephemeralKey: Base64URL.encode(ephemeral!.publicKey)
		))
	}

	private func handleFrame(_ frame: RemoteFrame) {
		switch frame {
		case let .hello(hello):
			if options.handshake == .accept, state != .online {
				Task { await self.handleInboundHello(hello) }
			} else {
				protocolViolation("unexpected hello")
			}
		case let .helloAck(ack):
			if options.handshake == .initiate { handleHelloAck(ack) } else { protocolViolation("unexpected hello_ack") }
		case let .pairingPending(frame):
			if options.handshake == .initiate { handlePairingPending(frame) } else { protocolViolation("unexpected pairing_pending") }
		case let .peerStatus(online):
			if !online { rejectPending("remote peer is offline") }
			emit(.peerStatus(online: online))
		case let .sealed(sealed):
			guard keys != nil else {
				if options.handshake == .initiate, isHandshaking, earlySealed.count < Self.earlySealedLimit {
					earlySealed.append(sealed)
					return
				}
				protocolViolation("sealed frame before handshake")
				return
			}
			openSealed(sealed)
		default:
			protocolViolation("plaintext \(frame.typeName) frame")
		}
	}

	private func handleInboundHello(_ hello: RemoteHello) async {
		guard let peerKey = try? RemoteCrypto.decodePublicKey(hello.identityKey, field: "identityKey") else {
			protocolViolation("hello identity key is invalid")
			return
		}
		if let expected = options.expectedPeerIdentityKey, !RemoteCrypto.bytesEqual(expected, peerKey) {
			fail(.unauthorized, "peer identity does not match the pinned key")
			return
		}
		peerIdentityKey = peerKey
		peerDeviceId = hello.deviceId
		let decision: RemoteHelloDecision
		if let onHello = options.onHello {
			decision = await onHello(hello)
		} else {
			decision = options.expectedPeerIdentityKey != nil ? .approve : .reject(reason: "peer is not paired")
		}
		if state == .closed { return }
		switch decision {
		case let .reject(reason):
			fail(.unauthorized, reason)
			return
		case let .pending(approval):
			setState(.pendingApproval)
			do {
				try transport.send(.pairingPending(RemotePairingPending(
					connectionId: hello.connectionId,
					peerDeviceId: options.deviceId,
					peerIdentityKey: Base64URL.encode(options.identity.publicKey)
				)))
			} catch {
				return
			}
			let approved = await approval()
			if state != .pendingApproval { return }
			if !approved {
				fail(.approvalRejected, "pairing was not approved")
				return
			}
		case .approve:
			break
		}
		guard let ephemeral else { return }
		do {
			keys = try RemoteCrypto.deriveSessionKeys(
				role: options.role,
				identity: options.identity,
				ephemeral: ephemeral,
				peerIdentityKey: peerKey,
				peerEphemeralKey: try RemoteCrypto.decodePublicKey(hello.ephemeralKey, field: "ephemeralKey")
			)
			try transport.send(.helloAck(RemoteHelloAck(
				connectionId: hello.connectionId,
				peerDeviceId: options.deviceId,
				peerIdentityKey: Base64URL.encode(options.identity.publicKey),
				peerEphemeralKey: Base64URL.encode(ephemeral.publicKey)
			)))
		} catch {
			protocolViolation(describe(error))
			return
		}
		goOnline()
	}

	private func handleHelloAck(_ frame: RemoteHelloAck) {
		// A stale acknowledgement from an earlier attempt must not make this transport appear online.
		guard frame.connectionId == connectionId else { return }
		guard let ephemeral else { return }
		guard let peerKey = try? RemoteCrypto.decodePublicKey(frame.peerIdentityKey, field: "peerIdentityKey") else {
			protocolViolation("hello_ack identity key is invalid")
			return
		}
		if let expected = options.expectedPeerIdentityKey, !RemoteCrypto.bytesEqual(expected, peerKey) {
			fail(.unauthorized, "peer identity does not match the pinned key")
			return
		}
		do {
			keys = try RemoteCrypto.deriveSessionKeys(
				role: options.role,
				identity: options.identity,
				ephemeral: ephemeral,
				peerIdentityKey: peerKey,
				peerEphemeralKey: try RemoteCrypto.decodePublicKey(frame.peerEphemeralKey, field: "peerEphemeralKey")
			)
		} catch {
			protocolViolation(describe(error))
			return
		}
		peerIdentityKey = peerKey
		peerDeviceId = frame.peerDeviceId
		goOnline()
	}

	private func handlePairingPending(_ frame: RemotePairingPending) {
		guard frame.connectionId == connectionId else { return }
		guard let peerKey = try? RemoteCrypto.decodePublicKey(frame.peerIdentityKey, field: "peerIdentityKey") else {
			protocolViolation("pairing_pending identity key is invalid")
			return
		}
		peerIdentityKey = peerKey
		if let expected = options.expectedPeerIdentityKey, !RemoteCrypto.bytesEqual(expected, peerKey) {
			fail(.unauthorized, "peer identity does not match the pinned key")
			return
		}
		peerDeviceId = frame.peerDeviceId
		setState(.pendingApproval)
	}

	private var isHandshaking: Bool {
		state == .connecting || state == .reconnecting || state == .pendingApproval
	}

	private func openSealed(_ sealed: RemoteSealed) {
		guard let keys else { return }
		let inner: RemoteFrame
		do {
			inner = try RemoteCrypto.openFrame(key: keys.receiveKey, sealed: sealed)
		} catch {
			protocolViolation(describe(error))
			return
		}
		handleSessionFrame(inner)
	}

	private func goOnline() {
		setState(.online)
		let early = earlySealed
		earlySealed.removeAll()
		for sealed in early {
			if state != .online { return }
			openSealed(sealed)
		}
		// Tell the peer what we already hold so it replays only the missing tail.
		try? sendSealed(.resume(lastEventSequence: lastEventSequence))
	}

	private func handleSessionFrame(_ frame: RemoteFrame) {
		switch frame {
		case let .request(request):
			emit(.remoteRequest(request))
		case let .response(response):
			handleResponse(response)
		case let .event(event):
			handleEvent(event)
		case let .ack(sequence):
			lastAckSequence = max(lastAckSequence, sequence)
			journal.acknowledge(sequence)
		case let .resume(lastEventSequence):
			replay(after: lastEventSequence)
		default:
			break
		}
	}

	private func replay(after sequence: Int) {
		guard let events = journal.replay(after: sequence) else {
			// The tail was evicted; the peer must reload state instead.
			_ = try? emitEvent(.sessionResync)
			return
		}
		for event in events {
			do { try sendSealed(.event(event)) } catch { return }
		}
	}

	private func handleResponse(_ frame: RemoteResponse) {
		guard let entry = pending.removeValue(forKey: frame.requestId) else { return }
		entry.timeout.cancel()
		lastRttMs = max(0, options.now() - entry.startedAt)
		if frame.success {
			entry.continuation.resume(returning: frame.payload)
		} else {
			lastErrorCode = frame.error?.code ?? .internalError
			let message = frame.error?.message ?? "remote request failed"
			emit(.error(frame.error ?? RemoteError(code: .internalError, message: message, retryable: false)))
			entry.continuation.resume(throwing: RemoteRequestError(message))
		}
	}

	private func handleEvent(_ event: RemoteEvent) {
		if event.name == .sessionResync {
			// An explicit reset: adopt the peer's sequence and let the application reload.
			lastEventSequence = event.sequence
			if state == .recovering { setState(.online) }
			emit(.remoteEvent(event))
			acknowledge(event.sequence)
			return
		}
		if event.sequence <= lastEventSequence { return }
		if event.sequence != lastEventSequence + 1 {
			lastErrorCode = .transportClosed
			setState(.recovering)
			try? sendSealed(.resume(lastEventSequence: lastEventSequence))
			return
		}
		lastEventSequence = event.sequence
		if state == .recovering { setState(.online) }
		emit(.remoteEvent(event))
		acknowledge(event.sequence)
	}

	private func acknowledge(_ sequence: Int) {
		try? sendSealed(.ack(sequence: sequence))
	}

	private func sendSealed(_ frame: RemoteFrame) throws {
		guard let keys else { throw RemoteRequestError("remote connection has no session keys") }
		try transport.send(.sealed(RemoteCrypto.sealFrame(key: keys.sendKey, frame: frame, randomBytes: options.randomBytes)))
	}

	private func protocolViolation(_ reason: String) {
		fail(.invalidFrame, reason)
	}

	private func fail(_ code: RemoteErrorCode, _ message: String) {
		lastErrorCode = code
		emit(.error(RemoteError(code: code, message: message, retryable: code == .invalidFrame)))
		rejectPending(message)
		keys = nil
		setState(.failed)
		transport.close(reason: message)
	}

	private func handleClose(_ reason: String?) {
		if state == .closed { return }
		reconnectCount += 1
		keys = nil
		rejectPending("remote transport closed")
		if state != .failed { setState(.reconnecting) }
	}

	private func rejectPending(_ message: String) {
		if !pending.isEmpty { lastErrorCode = .transportClosed }
		let entries = pending.values
		pending.removeAll()
		for entry in entries {
			entry.timeout.cancel()
			entry.continuation.resume(throwing: RemoteRequestError(message))
		}
	}

	private func setState(_ next: RemoteConnectionState) {
		if state == next { return }
		state = next
		emit(.state(next))
	}

	private func emit(_ event: RemoteConnectionEvent) {
		listeners.emit(event)
	}

	private func describe(_ error: Error) -> String {
		if let error = error as? RemoteProtocolError { return error.message }
		return (error as? LocalizedError)?.errorDescription ?? String(describing: error)
	}
}

File: apps/mobile/client-android/shared/src/commonMain/kotlin/org/vetta/android/domain/remote/connection/RemoteConnection.kt
package org.vetta.android.domain.remote.connection

import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancelAndJoin
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withTimeout
import kotlinx.serialization.json.JsonElement
import org.vetta.android.domain.remote.protocol.REMOTE_PROTOCOL_VERSION
import org.vetta.android.domain.remote.protocol.RemoteAck
import org.vetta.android.domain.remote.protocol.RemoteCapabilities
import org.vetta.android.domain.remote.protocol.RemoteCrypto
import org.vetta.android.domain.remote.protocol.RemoteError
import org.vetta.android.domain.remote.protocol.RemoteErrorCode
import org.vetta.android.domain.remote.protocol.RemoteEvent
import org.vetta.android.domain.remote.protocol.RemoteEventName
import org.vetta.android.domain.remote.protocol.RemoteFrame
import org.vetta.android.domain.remote.protocol.RemoteHello
import org.vetta.android.domain.remote.protocol.RemoteHelloAck
import org.vetta.android.domain.remote.protocol.RemoteIdentityKeyPair
import org.vetta.android.domain.remote.protocol.RemotePairingPending
import org.vetta.android.domain.remote.protocol.RemotePeerStatus
import org.vetta.android.domain.remote.protocol.RemoteRequest
import org.vetta.android.domain.remote.protocol.RemoteRequestMethod
import org.vetta.android.domain.remote.protocol.RemoteResponse
import org.vetta.android.domain.remote.protocol.RemoteResume
import org.vetta.android.domain.remote.protocol.RemoteRole
import org.vetta.android.domain.remote.protocol.RemoteSealed
import org.vetta.android.domain.remote.protocol.RemoteSessionFrame
import org.vetta.android.domain.remote.protocol.RemoteSessionKeys

enum class RemoteConnectionState {
    Idle,
    Connecting,
    PendingApproval,
    Online,
    Recovering,
    Reconnecting,
    Closed,
    Failed,
}

sealed interface RemoteConnectionEvent {
    data class RequestReceived(val request: RemoteRequest) : RemoteConnectionEvent

    data class EventReceived(val event: RemoteEvent) : RemoteConnectionEvent

    data class PeerStatusChanged(val online: Boolean) : RemoteConnectionEvent

    data class ErrorReceived(val error: RemoteError) : RemoteConnectionEvent
}

data class RemoteConnectionSnapshot(
    val state: RemoteConnectionState,
    val deviceId: String,
    val connectionId: String,
    val peerDeviceId: String?,
    val peerIdentityKey: String?,
    val lastEventSequence: Long,
    val lastAckSequence: Long,
    val pendingRequestCount: Int,
    val reconnectCount: Int,
    val lastRttMs: Long?,
    val lastErrorCode: RemoteErrorCode?,
)

data class RemoteConnectionOptions(
    val role: RemoteRole,
    val deviceId: String,
    val deviceName: String,
    val capabilities: RemoteCapabilities,
    val identity: RemoteIdentityKeyPair,
    /**
     * The desktop's pinned identity key. Null only for a manual pairing, where the
     * key is learnt from the desktop and the person compares the verification code.
     */
    val expectedPeerIdentityKey: ByteArray?,
    val connectionId: String,
    val requestTimeoutMs: Long = 30_000,
    /** The last event this phone saw from the desktop; the handshake resumes after it. */
    val resumeFrom: Long = 0,
    val randomBytes: (Int) -> ByteArray = RemoteCrypto::randomBytes,
)

class RemoteRequestException(val remoteError: RemoteError) : IllegalStateException(remoteError.message)

class RemoteConnection(
    private val transport: RemoteTransport,
    private val options: RemoteConnectionOptions,
    private val scope: CoroutineScope,
    private val logger: RemoteLogger = NoopRemoteLogger,
    private val now: () -> Long,
) {
    private data class PendingRequest(
        val startedAt: Long,
        val result: CompletableDeferred<JsonElement?>,
    )

    private val mutex = Mutex()
    private val pending = mutableMapOf<String, PendingRequest>()
    private val _state = MutableStateFlow(RemoteConnectionState.Idle)
    private val _events = MutableSharedFlow<RemoteConnectionEvent>(extraBufferCapacity = 32)
    private val earlySealed = mutableListOf<RemoteSealed>()
    private var incomingJob: Job? = null
    private var requestCounter = 0L
    private var ephemeral: RemoteIdentityKeyPair? = null
    private var keys: RemoteSessionKeys? = null
    private var peerDeviceId: String? = null
    private var peerIdentityKey: ByteArray? = null
    private var lastEventSequence = options.resumeFrom
    private var lastAckSequence = 0L
    private var reconnectCount = 0
    private var lastRttMs: Long? = null
    private var lastErrorCode: RemoteErrorCode? = null

    val state: StateFlow<RemoteConnectionState> = _state.asStateFlow()
    val events: SharedFlow<RemoteConnectionEvent> = _events.asSharedFlow()

    suspend fun connect() {
        if (_state.value == RemoteConnectionState.Online || _state.value == RemoteConnectionState.Connecting) return
        _state.value = if (_state.value == RemoteConnectionState.Idle) RemoteConnectionState.Connecting else RemoteConnectionState.Reconnecting
        ephemeral = RemoteCrypto.generateIdentityKeyPair(options.randomBytes)
        keys = null
        earlySealed.clear()
        try {
            incomingJob?.cancelAndJoin()
            incomingJob =
                scope.launch {
                    try {
                        transport.incoming.collect(::handleFrame)
                        handleTransportClosed("incoming stream completed")
                    } catch (error: Throwable) {
                        handleTransportClosed(error.message)
                    }
                }
            transport.connect()
            val currentEphemeral = requireNotNull(ephemeral)
            transport.send(
                RemoteHello(
                    protocolVersion = REMOTE_PROTOCOL_VERSION,
                    role = options.role,
                    deviceId = options.deviceId,
                    deviceName = options.deviceName,
                    capabilities = options.capabilities,
                    connectionId = options.connectionId,
                    identityKey = RemoteCrypto.toBase64Url(options.identity.publicKey),
                    ephemeralKey = RemoteCrypto.toBase64Url(currentEphemeral.publicKey),
                ),
            )
            logger.info(
                "remote connection handshake sent",
                mapOf("deviceId" to options.deviceId, "connectionId" to options.connectionId),
            )
        } catch (error: Throwable) {
            _state.value = RemoteConnectionState.Failed
            throw error
        }
    }

    suspend fun close() {
        if (_state.value == RemoteConnectionState.Closed) return
        _state.value = RemoteConnectionState.Closed
        keys = null
        earlySealed.clear()
        rejectPending(RemoteErrorCode.TransportClosed, "Remote connection closed")
        incomingJob?.cancelAndJoin()
        incomingJob = null
        transport.close()
    }

    suspend fun request(
        method: RemoteRequestMethod,
        payload: JsonElement? = null,
        sessionId: String? = null,
    ): JsonElement? {
        requireOnline()
        val requestId = "${options.connectionId}-${now()}-${requestCounter++}"
        val result = CompletableDeferred<JsonElement?>()
        mutex.withLock { pending[requestId] = PendingRequest(startedAt = now(), result = result) }
        try {
            sendSealed(RemoteRequest(requestId = requestId, method = method, sessionId = sessionId, payload = payload))
            return withTimeout(options.requestTimeoutMs) { result.await() }
        } catch (error: kotlinx.coroutines.TimeoutCancellationException) {
            mutex.withLock { pending.remove(requestId) }
            lastErrorCode = RemoteErrorCode.RequestTimeout
            val remoteError = RemoteError(RemoteErrorCode.RequestTimeout, "Remote request timed out", retryable = true)
            _events.emit(RemoteConnectionEvent.ErrorReceived(remoteError))
            throw RemoteRequestException(remoteError)
        } catch (error: Throwable) {
            mutex.withLock { pending.remove(requestId) }
            throw error
        }
    }

    suspend fun respond(requestId: String, payload: JsonElement?) {
        requireOnline()
        sendSealed(RemoteResponse(requestId = requestId, success = true, payload = payload))
    }

    suspend fun respond(requestId: String, error: RemoteError) {
        requireOnline()
        sendSealed(RemoteResponse(requestId = requestId, success = false, error = error))
    }

    suspend fun snapshot(): RemoteConnectionSnapshot =
        mutex.withLock {
            RemoteConnectionSnapshot(
                state = _state.value,
                deviceId = options.deviceId,
                connectionId = options.connectionId,
                peerDeviceId = peerDeviceId,
                peerIdentityKey = peerIdentityKey?.let(RemoteCrypto::toBase64Url),
                lastEventSequence = lastEventSequence,
                lastAckSequence = lastAckSequence,
                pendingRequestCount = pending.size,
                reconnectCount = reconnectCount,
                lastRttMs = lastRttMs,
                lastErrorCode = lastErrorCode,
            )
        }

    private suspend fun handleFrame(frame: RemoteFrame) {
        when (frame) {
            is RemoteHelloAck -> handleHelloAck(frame)
            is RemotePairingPending -> handlePairingPending(frame)
            is RemotePeerStatus -> {
                if (!frame.online) rejectPending(RemoteErrorCode.TransportClosed, "Remote peer is offline")
                _events.emit(RemoteConnectionEvent.PeerStatusChanged(frame.online))
            }
            is RemoteSealed -> {
                if (keys == null && isHandshaking() && earlySealed.size < EARLY_SEALED_LIMIT) {
                    earlySealed += frame
                } else {
                    handleSealed(frame)
                }
            }
            is RemoteHello,
            is RemoteRequest,
            is RemoteResponse,
            is RemoteEvent,
            is RemoteAck,
            is RemoteResume,
            -> protocolFailure("Unexpected plaintext frame after protocol v2 handshake")
        }
    }

    private suspend fun handleHelloAck(frame: RemoteHelloAck) {
        if (frame.connectionId != options.connectionId) return
        val currentEphemeral = ephemeral ?: return
        val peerIdentity = RemoteCrypto.decodePublicKey(frame.peerIdentityKey, "peerIdentityKey")
        if (!matchesExpectedPeer(peerIdentity)) {
            protocolFailure("Peer identity does not match the paired desktop")
            return
        }
        keys =
            RemoteCrypto.deriveSessionKeys(
                role = options.role,
                identity = options.identity,
                ephemeral = currentEphemeral,
                peerIdentityKey = peerIdentity,
                peerEphemeralKey = RemoteCrypto.decodePublicKey(frame.peerEphemeralKey, "peerEphemeralKey"),
            )
        peerIdentityKey = peerIdentity
        peerDeviceId = frame.peerDeviceId
        _state.value = RemoteConnectionState.Online
        logger.info(
            "remote connection online",
            mapOf("peerDeviceId" to frame.peerDeviceId, "connectionId" to options.connectionId),
        )
        val queued = earlySealed.toList()
        earlySealed.clear()
        for (sealed in queued) handleSealed(sealed)
        sendSealed(RemoteResume(lastEventSequence = lastEventSequence))
    }

    private suspend fun handlePairingPending(frame: RemotePairingPending) {
        if (frame.connectionId != options.connectionId) return
        val peerIdentity = RemoteCrypto.decodePublicKey(frame.peerIdentityKey, "peerIdentityKey")
        if (!matchesExpectedPeer(peerIdentity)) {
            protocolFailure("Pairing response came from an unexpected desktop")
            return
        }
        peerIdentityKey = peerIdentity
        peerDeviceId = frame.peerDeviceId
        _state.value = RemoteConnectionState.PendingApproval
    }

    private fun matchesExpectedPeer(peerIdentity: ByteArray): Boolean =
        options.expectedPeerIdentityKey?.let { RemoteCrypto.bytesEqual(it, peerIdentity) } ?: true

    private suspend fun handleSealed(frame: RemoteSealed) {
        val sessionKeys = keys
        if (sessionKeys == null) {
            protocolFailure("Encrypted frame arrived before the handshake completed")
            return
        }
        val inner =
            try {
                RemoteCrypto.openFrame(sessionKeys.receiveKey, frame)
            } catch (error: Throwable) {
                protocolFailure(error.message ?: "Encrypted frame could not be opened")
                return
            }
        handleSessionFrame(inner)
    }

    private suspend fun handleSessionFrame(frame: RemoteSessionFrame) {
        when (frame) {
            is RemoteRequest -> _events.emit(RemoteConnectionEvent.RequestReceived(frame))
            is RemoteResponse -> handleResponse(frame)
            is RemoteEvent -> handleEvent(frame)
            is RemoteAck -> lastAckSequence = maxOf(lastAckSequence, frame.sequence)
            is RemoteResume -> logger.debug("remote resume received", mapOf("sequence" to frame.lastEventSequence))
        }
    }

    private suspend fun handleResponse(frame: RemoteResponse) {
        val request = mutex.withLock { pending.remove(frame.requestId) }
        if (request == null) {
            logger.warn("remote response has no pending request", mapOf("requestId" to frame.requestId))
            return
        }
        lastRttMs = maxOf(0, now() - request.startedAt)
        if (frame.success) {
            request.result.complete(frame.payload)
        } else {
            val error = frame.error ?: RemoteError(RemoteErrorCode.InternalError, "Remote request failed", false)
            lastErrorCode = error.code
            _events.emit(RemoteConnectionEvent.ErrorReceived(error))
            request.result.completeExceptionally(RemoteRequestException(error))
        }
    }

    private suspend fun handleEvent(event: RemoteEvent) {
        if (event.name == RemoteEventName.SessionResync) {
            lastEventSequence = event.sequence
            if (_state.value == RemoteConnectionState.Recovering) _state.value = RemoteConnectionState.Online
            _events.emit(RemoteConnectionEvent.EventReceived(event))
            sendSealed(RemoteAck(sequence = event.sequence))
            return
        }
        if (event.sequence <= lastEventSequence) {
            logger.debug("duplicate remote event ignored", mapOf("sequence" to event.sequence, "eventId" to event.eventId))
            return
        }
        if (event.sequence != lastEventSequence + 1) {
            lastErrorCode = RemoteErrorCode.TransportClosed
            _state.value = RemoteConnectionState.Recovering
            logger.warn(
                "remote event sequence gap",
                mapOf("expected" to lastEventSequence + 1, "received" to event.sequence),
            )
            sendSealed(RemoteResume(lastEventSequence = lastEventSequence))
            return
        }
        lastEventSequence = event.sequence
        if (_state.value == RemoteConnectionState.Recovering) _state.value = RemoteConnectionState.Online
        _events.emit(RemoteConnectionEvent.EventReceived(event))
        sendSealed(RemoteAck(sequence = event.sequence))
    }

    private suspend fun sendSealed(frame: RemoteSessionFrame) {
        val sessionKeys = keys ?: error("Remote connection has no session keys")
        transport.send(RemoteCrypto.sealFrame(sessionKeys.sendKey, frame, randomBytes = options.randomBytes))
    }

    private suspend fun protocolFailure(message: String) {
        lastErrorCode = RemoteErrorCode.InvalidFrame
        val error = RemoteError(RemoteErrorCode.InvalidFrame, message, retryable = false)
        _events.emit(RemoteConnectionEvent.ErrorReceived(error))
        rejectPending(RemoteErrorCode.InvalidFrame, message)
        keys = null
        _state.value = RemoteConnectionState.Failed
        transport.close()
    }

    private suspend fun handleTransportClosed(reason: String?) {
        if (_state.value == RemoteConnectionState.Closed || _state.value == RemoteConnectionState.Failed) return
        reconnectCount += 1
        keys = null
        rejectPending(RemoteErrorCode.TransportClosed, "Remote transport closed")
        logger.warn(
            "remote transport closed",
            mapOf(
                "deviceId" to options.deviceId,
                "connectionId" to options.connectionId,
                "reason" to reason,
                "reconnectCount" to reconnectCount,
            ),
        )
        _state.value = RemoteConnectionState.Reconnecting
    }

    private suspend fun rejectPending(code: RemoteErrorCode, message: String) {
        lastErrorCode = code
        val requests = mutex.withLock { pending.values.toList().also { pending.clear() } }
        val error = RemoteRequestException(RemoteError(code, message, retryable = true))
        requests.forEach { it.result.completeExceptionally(error) }
    }

    private fun requireOnline() {
        check(_state.value == RemoteConnectionState.Online) {
            "Remote connection is ${_state.value.name.lowercase()}"
        }
    }

    private fun isHandshaking(): Boolean =
        _state.value == RemoteConnectionState.Connecting ||
            _state.value == RemoteConnectionState.Reconnecting ||
            _state.value == RemoteConnectionState.PendingApproval

    private companion object {
        const val EARLY_SEALED_LIMIT = 32
    }
}
